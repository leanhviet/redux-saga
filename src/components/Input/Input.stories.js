// Libs
import React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'

// Components
import Input from './index'

storiesOf('Input', module)
  .add('Default', () => <Input />)
  .add('Input has handle event on enter', () => (
    <Input
      placeholder="TextBox"
      defaultValue="Viet"
      handleOnKeyDown={action('entered')}
    />
  ))
