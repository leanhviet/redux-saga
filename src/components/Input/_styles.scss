.input {
  border: 1px solid $iron-approx;
  color: $pale-sky-approx;
  font-size: 16px;
  padding: 17px 25px;
  width: 100%;
  outline: none;

  &.search {
    background: url('./assets/images/search.png') $white no-repeat right 35px
      center;
  }
}
