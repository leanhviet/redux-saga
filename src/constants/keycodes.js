export default {
  KEY_ENTER: 13
}
