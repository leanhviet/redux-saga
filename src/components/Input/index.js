// @flow
import React, { createRef, useState } from 'react'

// Constants
import keycodes from '../../constants/keycodes'

type Props = {
  type: string,
  defaultValue: string,
  placeholder: string,
  innerRef: Object,
  keyDown: number,
  handleOnKeyDown: Function,
  className: string
}

const Input = (props: Props) => {
  const {
    type,
    defaultValue,
    placeholder,
    innerRef,
    keyDown,
    handleOnKeyDown,
    className
  } = props

  const [valueInput, setValueInput] = useState(defaultValue)

  // handle event on enter
  const handleKeyDown = event => {
    const { keyCode, target } = event
    const valueFromKeyDown = target.value

    if (keyCode === keyDown) {
      handleOnKeyDown(valueFromKeyDown)
    }
  }

  return (
    <input
      type={type}
      className={`input ${className}`}
      value={valueInput}
      placeholder={placeholder}
      onChange={event => setValueInput(event.target.value)}
      onKeyDown={handleKeyDown}
      ref={innerRef}
    />
  )
}

Input.defaultProps = {
  type: 'text',
  defaultValue: '',
  placeholder: '',
  innerRef: createRef(),
  keyDown: keycodes.KEY_ENTER,
  handleOnKeyDown: () => {},
  className: ''
}

export default Input
