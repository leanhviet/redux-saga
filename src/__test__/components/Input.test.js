// Libs
import React from 'react'
import { shallowToJson } from 'enzyme-to-json'
import { shallow } from 'enzyme'

// Component
import Input from '../../components/Input'

// Constants
import keycodes from '../../constants/keycodes'

const mockKeyDownFunc = jest.fn()
const mockProps = {
  type: 'text',
  defaultValue: 'Robert Downey Jr',
  placeholder: 'Enter on textbox, please!',
  className: 'mockClass',
  handleOnKeyDown: mockKeyDownFunc
}

const wrapper = (props = {}) => {
  const component = shallow(<Input {...props} />)

  return component
}

describe('Input component', () => {
  const wrapperWithoutProps = wrapper()

  // Snapshot
  test('should render correctly Input component default props', () => {
    expect(shallowToJson(wrapperWithoutProps)).toMatchSnapshot()
  })

  // Testing props
  describe('Handle with props', () => {
    test('should is rendered default props', () => {
      expect(wrapperWithoutProps.find('input').props().defaultValue).toBeFalsy()
    })

    test('should is rendered with a custom Props', () => {
      const wrapperWithProps = wrapper(mockProps)

      // Expected with a placeholder
      expect(wrapperWithProps.find('input').props().placeholder).toMatch(
        mockProps.placeholder
      )

      // Expected with a defaultValue
      expect(wrapperWithProps.find('input').props().value).toMatch(
        mockProps.defaultValue
      )

      // Expected with a defaultValue
      expect(wrapperWithProps.find('input').props().className).toMatch(
        mockProps.className
      )
    })
  })

  // Testing events
  describe('Checking events', () => {
    test('check the on Enter Press callback ', () => {
      const InputComponent = wrapper(mockProps)

      const input = InputComponent.find('input')
      const value = 'John Hart J'

      input.simulate('keydown', {
        keyCode: keycodes.KEY_ENTER,
        target: { value }
      })

      expect(mockKeyDownFunc).toHaveBeenCalledTimes(1)
    })

    test('check the on Change callback ', () => {
      const input = wrapperWithoutProps.find('input')
      const value = 'John Hart Jr'

      input.simulate('change', {
        target: { value }
      })

      expect(wrapperWithoutProps.find('input').props().value).toEqual(value)
    })
  })
})
