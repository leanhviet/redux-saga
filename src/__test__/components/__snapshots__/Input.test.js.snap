// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component should render correctly Input component default props 1`] = `
<input
  className="input "
  onChange={[Function]}
  onKeyDown={[Function]}
  placeholder=""
  type="text"
  value=""
/>
`;
